### PR TITLE
masync: add hashmap example

### PR DESCRIPTION
**Pull request by kswiecicki**
This change is [Reviewable](https://reviewable.io/reviews/pmem/miniasync/71)

### CODE REVIEW

**Review by pbalcer — CHANGES_REQUESTED** [review]
> Reviewed 1 of 2 files at r1.
> *[Reviewable](https://reviewable.io/reviews/pmem/miniasync/71)* status: 1 of 2 files reviewed, 5 unresolved discussions (waiting on @kswiecicki and @pbalcer)

___
*[examples/hashmap/hashmap.c, line 15 at r1](https://reviewable.io/reviews/pmem/miniasync/71#-MyqzUvM-grCKPgauSk7:-MyqzUvM-grCKPgauSk8:bcbbd90) ([raw file](https://github.com/pmem/miniasync/blob/9915719fb5ca0b063ef2e5d1b3b9708c689ee20b/examples/hashmap/hashmap.c#L15)):*
> ```C
> 
> /* Polls 'nfuts' number of futures until they complete, makes us of runtime */
> #define WAIT_FUTURES(runtimep, futsp, nfuts)\
> ```

hm, if you find yourself needing this often, maybe it would be better to create a variadic variant of runtime_wait (i.e., `runtime_waitv(&a, &b, &c, RUNTIME_END);` or something like that)
___
*[examples/hashmap/hashmap.c, line 27 at r1](https://reviewable.io/reviews/pmem/miniasync/71#-Myr0XhVAlr4u7WqHyx5:-Myr0XhW5374PMx9flMs:b-1fs9il) ([raw file](https://github.com/pmem/miniasync/blob/9915719fb5ca0b063ef2e5d1b3b9708c689ee20b/examples/hashmap/hashmap.c#L27)):*
> ```C
> struct hashmap_entry {
> 	uint64_t key;
> 	char value[HASHMAP_VALUE_SIZE];
> ```

I thought we'd allocate values dynamically on put? This way there's no need for a size limit.
___
*[examples/hashmap/hashmap.c, line 35 at r1](https://reviewable.io/reviews/pmem/miniasync/71#-Myr1se8CBEpIJ_PAuA0:-Myr1se8CBEpIJ_PAuA1:b-o0uq0e) ([raw file](https://github.com/pmem/miniasync/blob/9915719fb5ca0b063ef2e5d1b3b9708c689ee20b/examples/hashmap/hashmap.c#L35)):*
> ```C
> 	size_t capacity; /* max stored entries */
> 	size_t length; /* number of stored entries */
> 	unsigned lock; /* used for hashmap operations synchronization */
> ```

I think you can have per-entry locks (you could probably reuse the `deleted` variable, so 0 - empty, 1 - being written to, 2 - deleted or something). The only global state that is modified by normal operations is 'length', which you can safely modify using atomics (fetch and add, fetch and sub).
___
*[examples/hashmap/hashmap.c, line 62 at r1](https://reviewable.io/reviews/pmem/miniasync/71#-Myr-JyH4GqEL7PUeOux:-Myr-JyH4GqEL7PUeOuy:b-97nirk) ([raw file](https://github.com/pmem/miniasync/blob/9915719fb5ca0b063ef2e5d1b3b9708c689ee20b/examples/hashmap/hashmap.c#L62)):*
> ```C
> hashmap_new(size_t capacity)
> {
> 	if (!is_power_of_2(capacity)) {
> ```

we don't really have to be this performance conscious in examples. just use a simple modulo for the hash function.
___
*[examples/hashmap/hashmap.c, line 238 at r1](https://reviewable.io/reviews/pmem/miniasync/71#-Myr069mBYBcnJxNIhef:-Myr069mBYBcnJxNIheg:b-wsymof) ([raw file](https://github.com/pmem/miniasync/blob/9915719fb5ca0b063ef2e5d1b3b9708c689ee20b/examples/hashmap/hashmap.c#L238)):*
> ```C
> 
> struct hashmap_get_output {
> 	char *value;
> ```

it would probably be better (safer and more realistic) to provide a copy of the value. it would also allow you to showcase DSA more because right now get doesn't really need to be async ;)

**Comment by kswiecicki** [comment]
___
*[examples/hashmap/hashmap.c, line 62 at r1](https://reviewable.io/reviews/pmem/miniasync/71#-Myr-JyH4GqEL7PUeOux:-MzG2-jnA5v6ODonWJDk:b-896fix) ([raw file](https://github.com/pmem/miniasync/blob/9915719fb5ca0b063ef2e5d1b3b9708c689ee20b/examples/hashmap/hashmap.c#L62)):*
<details><summary><i>Previously, pbalcer (Piotr Balcer) wrote…</i></summary><blockquote>

we don't really have to be this performance conscious in examples. just use a simple modulo for the hash function.
</blockquote></details>

Done.

**Comment by kswiecicki** [comment]
___
*[examples/hashmap/hashmap.c, line 238 at r1](https://reviewable.io/reviews/pmem/miniasync/71#-Myr069mBYBcnJxNIhef:-MzG22_l5i-oKuCEACCn:b-896fix) ([raw file](https://github.com/pmem/miniasync/blob/9915719fb5ca0b063ef2e5d1b3b9708c689ee20b/examples/hashmap/hashmap.c#L238)):*
<details><summary><i>Previously, pbalcer (Piotr Balcer) wrote…</i></summary><blockquote>

it would probably be better (safer and more realistic) to provide a copy of the value. it would also allow you to showcase DSA more because right now get doesn't really need to be async ;)
</blockquote></details>

Done.

**Review by kswiecicki — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/pmem/miniasync/71)* status: 1 of 3 files reviewed, 7 unresolved discussions (waiting on @kswiecicki, @lukaszstolarczuk, and @pbalcer)

*[a discussion](https://reviewable.io/reviews/pmem/miniasync/71#-Mz0ixx1DOY83gt9Mzfb:-MzG461d3hA6wK2hjvQD:b-896fix) (no related file):*
<details><summary><i>Previously, lukaszstolarczuk (Łukasz Stolarczuk) wrote…</i></summary><blockquote>

you could add this example to execution in tests
</blockquote></details>

Done.

---
___
*[examples/hashmap/hashmap.c, line 645 at r2](https://reviewable.io/reviews/pmem/miniasync/71#-Mz0ihYTCwNKuo88ngvU:-MzG3qd4BQA1hHjwa5CP:b-896fix) ([raw file](https://github.com/pmem/miniasync/blob/6d3d965d4cf221a9095aeeaaa7eca2f2c6fd2d66/examples/hashmap/hashmap.c#L645)):*
<details><summary><i>Previously, lukaszstolarczuk (Łukasz Stolarczuk) wrote…</i></summary><blockquote>

`entries` (misspell)
</blockquote></details>

Done.

**Review by lukaszstolarczuk — COMMENTED** [review]
> Reviewed 1 of 2 files at r1, 1 of 2 files at r3, 1 of 1 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/pmem/miniasync/71)* status: 3 of 4 files reviewed, 5 unresolved discussions (waiting on @kswiecicki, @lukaszstolarczuk, and @pbalcer)

**Review by pbalcer — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/pmem/miniasync/71)* status: 2 of 4 files reviewed, 8 unresolved discussions (waiting on @kswiecicki, @lukaszstolarczuk, and @pbalcer)

___
*[`examples/hashmap/hashmap.c` line 49 at r7](https://reviewable.io/reviews/pmem/miniasync/71#-N0VNP-19BC-kIjoAnDV:-N0VNP-19BC-kIjoAnDW:b-hqvmj6) ([raw file](https://github.com/pmem/miniasync/blob/9508f0e10dc38b121f3dc7118782c2bec0f4139e/examples/hashmap/hashmap.c#L49)):*
> ```C
> enum hashmap_entry_state {
> 	HASHMAP_ENTRY_STATE_UNOCCUPIED,
> 	HASHMAP_ENTRY_STATE_PROCESSED,
> ```

PROCESSING? LOCKED? I found 'processed' to be misleading.
___
*[`examples/hashmap/hashmap.c` line 153 at r7](https://reviewable.io/reviews/pmem/miniasync/71#-N0VIJWB2m7dlM2HiFF-:-N0VIJWB2m7dlM2HiFF0:b-2d78mb) ([raw file](https://github.com/pmem/miniasync/blob/9508f0e10dc38b121f3dc7118782c2bec0f4139e/examples/hashmap/hashmap.c#L153)):*
> ```C
>  * 'HASHMAP_ENTRY_STATE_UNOCCUPIED' - looks for unoccupied entry
>  */
> static int
> ```

capacity is size_t, so size_t ? or ssize_t if you want to return an error (then change the capacity too, will help avoid casts).
___
*[`examples/hashmap/hashmap.c` line 161 at r7](https://reviewable.io/reviews/pmem/miniasync/71#-N0VI1qQ2LJ2FkXXL4RT:-N0VI1qQ2LJ2FkXXL4RU:bsfim6h) ([raw file](https://github.com/pmem/miniasync/blob/9508f0e10dc38b121f3dc7118782c2bec0f4139e/examples/hashmap/hashmap.c#L161)):*
> ```C
> 	switch (state) {
> 		case HASHMAP_ENTRY_STATE_UNOCCUPIED:
> 			for (size_t i = 0; i < hm->capacity; i++, index =
> ```

You can invert this.

for (...) {
  switch(state) {
  };
}

saves you from having to have two loops.
___
*[`examples/hashmap/hashmap.c` line 439 at r7](https://reviewable.io/reviews/pmem/miniasync/71#-N0VKUSq1Yliozir4upX:-N0VKUSq1Yliozir4upY:b9l82az) ([raw file](https://github.com/pmem/miniasync/blob/9508f0e10dc38b121f3dc7118782c2bec0f4139e/examples/hashmap/hashmap.c#L439)):*
> ```C
> 
> 	void *src = NULL;
> 	void *dest = NULL;
> ```

The destination buffer should be user-supplied, and the function should return the size copied in addition to the size of the entry.
___
*[`examples/hashmap/hashmap.c` line 483 at r7](https://reviewable.io/reviews/pmem/miniasync/71#-N0VL1v-8QEoMYn7N3WB:-N0VL1v-8QEoMYn7N3WC:b166zat) ([raw file](https://github.com/pmem/miniasync/blob/9508f0e10dc38b121f3dc7118782c2bec0f4139e/examples/hashmap/hashmap.c#L483)):*
> ```C
> 	struct hashmap_entry_set_state_fut fut = {.output.changed = 0};
> 	struct hashmap_entry *hme = data->lookup_lock_entry.fut.output.hme;
> 	if (hme == NULL) {
> ```

This will fail even in the case of "lock" contention, right? not just when the entry wasn't found. For example, two concurrent gets of the same value.
___
*[`examples/hashmap/hashmap.c` line 619 at r7](https://reviewable.io/reviews/pmem/miniasync/71#-N0VMbpy4TPIJEIzSqTR:-N0VMbpy4TPIJEIzSqTS:b6j0yw1) ([raw file](https://github.com/pmem/miniasync/blob/9508f0e10dc38b121f3dc7118782c2bec0f4139e/examples/hashmap/hashmap.c#L619)):*
> ```C
> 	struct hashmap_entry *hme = lookup_lock_output->hme;
> 
> 	if (hme == NULL) {
> ```

again, this will fail in the case of lock contention. Wouldn't it be better to retry the lookup?
___
*[`examples/hashmap/hashmap.c` line 1021 at r7](https://reviewable.io/reviews/pmem/miniasync/71#-N0VNinT7zKU_Q__K0_9:-N0VNinT7zKU_Q__K0_A:bcu8g9z) ([raw file](https://github.com/pmem/miniasync/blob/9508f0e10dc38b121f3dc7118782c2bec0f4139e/examples/hashmap/hashmap.c#L1021)):*
> ```C
> 	(void) put_output;
> 	(void) remove_output;
> ```

You are using atomic operations, but have you tested if this is mt-safe?

**Comment by kswiecicki** [comment]
*[`examples/hashmap/hashmap.c` line 153 at r7](https://reviewable.io/reviews/pmem/miniasync/71#-N0VIJWB2m7dlM2HiFF-:-N0_NkBO16D0vTBY1FAy:bep19b8) ([raw file](https://github.com/pmem/miniasync/blob/9508f0e10dc38b121f3dc7118782c2bec0f4139e/examples/hashmap/hashmap.c#L153)):*
<details><summary><i>Previously, pbalcer (Piotr Balcer) wrote…</i></summary><blockquote>

capacity is size_t, so size_t ? or ssize_t if you want to return an error (then change the capacity too, will help avoid casts).
</blockquote></details>

I switched the return to ssize_t, but by changing the `capacity` type I would need to introduce multiple of other casts. In `malloc` and `calloc` for example.

**Comment by kswiecicki** [comment]
*[`examples/hashmap/hashmap.c` line 483 at r7](https://reviewable.io/reviews/pmem/miniasync/71#-N0VL1v-8QEoMYn7N3WB:-N0_vlTu8-7--I3Mborz:b-n9ulf8) ([raw file](https://github.com/pmem/miniasync/blob/9508f0e10dc38b121f3dc7118782c2bec0f4139e/examples/hashmap/hashmap.c#L483)):*
<details><summary><i>Previously, pbalcer (Piotr Balcer) wrote…</i></summary><blockquote>

This will fail even in the case of "lock" contention, right? not just when the entry wasn't found. For example, two concurrent gets of the same value.
</blockquote></details>

I've changed the 'hashmap_lookup_lock_entry_fut' future to re-run its future entries in the case of lock contention.

**Comment by kswiecicki** [comment]
*[`examples/hashmap/hashmap.c` line 619 at r7](https://reviewable.io/reviews/pmem/miniasync/71#-N0VMbpy4TPIJEIzSqTR:-N0jFbkfAdEFIw8vMMDW:b-w2kpl9) ([raw file](https://github.com/pmem/miniasync/blob/9508f0e10dc38b121f3dc7118782c2bec0f4139e/examples/hashmap/hashmap.c#L619)):*
<details><summary><i>Previously, pbalcer (Piotr Balcer) wrote…</i></summary><blockquote>

again, this will fail in the case of lock contention. Wouldn't it be better to retry the lookup?
</blockquote></details>

I've added a future that re-runs provided chained future entries.
I'm not sure if such a use-case was intended in Miniasync, but it works.

**Review by kswiecicki — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/pmem/miniasync/71)* status: 2 of 4 files reviewed, 8 unresolved discussions (waiting on @kswiecicki, @lukaszstolarczuk, and @pbalcer)

___
*[`examples/hashmap/hashmap.c` line 49 at r7](https://reviewable.io/reviews/pmem/miniasync/71#-N0VNP-19BC-kIjoAnDV:-N0_1e6ED3stBba09zBc:b-896fix) ([raw file](https://github.com/pmem/miniasync/blob/9508f0e10dc38b121f3dc7118782c2bec0f4139e/examples/hashmap/hashmap.c#L49)):*
<details><summary><i>Previously, pbalcer (Piotr Balcer) wrote…</i></summary><blockquote>

PROCESSING? LOCKED? I found 'processed' to be misleading.
</blockquote></details>

Done.
___
*[`examples/hashmap/hashmap.c` line 161 at r7](https://reviewable.io/reviews/pmem/miniasync/71#-N0VI1qQ2LJ2FkXXL4RT:-N0_Njn26bF492tETNVF:b-896fix) ([raw file](https://github.com/pmem/miniasync/blob/9508f0e10dc38b121f3dc7118782c2bec0f4139e/examples/hashmap/hashmap.c#L161)):*
<details><summary><i>Previously, pbalcer (Piotr Balcer) wrote…</i></summary><blockquote>

You can invert this.

for (...) {
  switch(state) {
  };
}

saves you from having to have two loops.
</blockquote></details>

Done.
___
*[`examples/hashmap/hashmap.c` line 439 at r7](https://reviewable.io/reviews/pmem/miniasync/71#-N0VKUSq1Yliozir4upX:-N0_va_X8dCFYuVFABJJ:b-896fix) ([raw file](https://github.com/pmem/miniasync/blob/9508f0e10dc38b121f3dc7118782c2bec0f4139e/examples/hashmap/hashmap.c#L439)):*
<details><summary><i>Previously, pbalcer (Piotr Balcer) wrote…</i></summary><blockquote>

The destination buffer should be user-supplied, and the function should return the size copied in addition to the size of the entry.
</blockquote></details>

Done.

**Comment by kswiecicki** [comment]
*[`examples/hashmap/hashmap.c` line 1021 at r7](https://reviewable.io/reviews/pmem/miniasync/71#-N0VNinT7zKU_Q__K0_9:-N0kBthUEaGave2KNtrc:bh7h13q) ([raw file](https://github.com/pmem/miniasync/blob/9508f0e10dc38b121f3dc7118782c2bec0f4139e/examples/hashmap/hashmap.c#L1021)):*
<details><summary><i>Previously, pbalcer (Piotr Balcer) wrote…</i></summary><blockquote>

You are using atomic operations, but have you tested if this is mt-safe?
</blockquote></details>

No. I could spawn some threads and add them to the example, but I believe it could clutter the example.

**Review by pbalcer — CHANGES_REQUESTED** [review]
> Reviewed 1 of 2 files at r7, 1 of 1 files at r11.
> *[Reviewable](https://reviewable.io/reviews/pmem/miniasync/71)* status: all files reviewed (commit messages unreviewed), 5 unresolved discussions (waiting on @kswiecicki and @pbalcer)

___
*[`examples/hashmap/hashmap.c` line 1021 at r7](https://reviewable.io/reviews/pmem/miniasync/71#-N0VNinT7zKU_Q__K0_9:-N0ocfPq1f3GFW6wYkL4:b-mmmmji) ([raw file](https://github.com/pmem/miniasync/blob/9508f0e10dc38b121f3dc7118782c2bec0f4139e/examples/hashmap/hashmap.c#L1021)):*
<details><summary><i>Previously, kswiecicki (Krzysztof Święcicki) wrote…</i></summary><blockquote>

No. I could spawn some threads and add them to the example, but I believe it could clutter the example.
</blockquote></details>

Either remove the atomics or make sure it works. I don't think we need to add threads to this example, but we can't mislead people about thread safety.
___
*[`examples/hashmap/hashmap.c` line 191 at r11](https://reviewable.io/reviews/pmem/miniasync/71#-N0o_WcuCFbr9dgQTXqh:-N0o_WcuCFbr9dgQTXqi:byleaes) ([raw file](https://github.com/pmem/miniasync/blob/75168591949f098edb5dd35f30d3c21d6bdda74b/examples/hashmap/hashmap.c#L191)):*
> ```C
> 			case HASHMAP_ENTRY_STATE_PRESENT:
> 				if (hashmap_entry_deleted(hme)) {
> 				/* Deleted entries are skipped */
> ```

one more tab, same below
___
*[`examples/hashmap/hashmap.c` line 379 at r11](https://reviewable.io/reviews/pmem/miniasync/71#-N0o_uWZBh3sQjzwkvHo:-N0o_uWZBh3sQjzwkvHp:btc1roz) ([raw file](https://github.com/pmem/miniasync/blob/75168591949f098edb5dd35f30d3c21d6bdda74b/examples/hashmap/hashmap.c#L379)):*
> ```C
> 
> struct chain_entries_rerun_output {
> 	uint64_t empty; /* Avoid compiled empty struct error */
> ```

`uint64_t unused;`

**Review by kswiecicki — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/pmem/miniasync/71)* status: 3 of 4 files reviewed, 5 unresolved discussions (waiting on @kswiecicki and @pbalcer)

___
*[`examples/hashmap/hashmap.c` line 1021 at r7](https://reviewable.io/reviews/pmem/miniasync/71#-N0VNinT7zKU_Q__K0_9:-N1Dgcmz--RxOFSPDfBq:b-rjwa97) ([raw file](https://github.com/pmem/miniasync/blob/9508f0e10dc38b121f3dc7118782c2bec0f4139e/examples/hashmap/hashmap.c#L1021)):*
<details><summary><i>Previously, pbalcer (Piotr Balcer) wrote…</i></summary><blockquote>

Either remove the atomics or make sure it works. I don't think we need to add threads to this example, but we can't mislead people about thread safety.
</blockquote></details>

I've prepared and tested a multithreaded proof.
Take a look here: https://github.com/kswiecicki/libuasync/blob/masync-example-hashmap-threads/examples/hashmap/hashmap.c.
___
*[`examples/hashmap/hashmap.c` line 191 at r11](https://reviewable.io/reviews/pmem/miniasync/71#-N0o_WcuCFbr9dgQTXqh:-N1DgbMiBpjNE7s1jzoj:b-896fix) ([raw file](https://github.com/pmem/miniasync/blob/75168591949f098edb5dd35f30d3c21d6bdda74b/examples/hashmap/hashmap.c#L191)):*
<details><summary><i>Previously, pbalcer (Piotr Balcer) wrote…</i></summary><blockquote>

one more tab, same below
</blockquote></details>

Done.
___
*[`examples/hashmap/hashmap.c` line 379 at r11](https://reviewable.io/reviews/pmem/miniasync/71#-N0o_uWZBh3sQjzwkvHo:-N1Dgc5n3mH7abpg3lbS:b-896fix) ([raw file](https://github.com/pmem/miniasync/blob/75168591949f098edb5dd35f30d3c21d6bdda74b/examples/hashmap/hashmap.c#L379)):*
<details><summary><i>Previously, pbalcer (Piotr Balcer) wrote…</i></summary><blockquote>

`uint64_t unused;`
</blockquote></details>

Done.

File: examples/hashmap/hashmap.c [hunk 1]
```diff
@@ -0,0 +1,1268 @@
...
+#define WAIT_FUTURES_MAX 4
+
+/* Polls 'nfuts' number of futures until they complete, makes use of runtime */
+#define WAIT_FUTURES(runtimep, futsp, nfuts)\
+do {\
+	struct future *tmp_futs[WAIT_FUTURES_MAX];\
+	for (int i = 0; i < nfuts; i++) {\
+		tmp_futs[i] = FUTURE_AS_RUNNABLE(&(futsp[i]));\
+	}\
+\
```
Comment (DamianDuy) [inline examples/hashmap/hashmap.c:45]: Can those slashes be removed?

**Comment by kswiecicki** [comment]
*[`examples/hashmap/hashmap.c` line 45 at r12](https://reviewable.io/reviews/pmem/miniasync/71#-MymC2OhAVwVsyL67KRU-r12-45:-N1bytW-9uZ0rRJu0iok:b-xrt5k2) ([raw file](https://github.com/pmem/miniasync/blob/f8dfa4efa4a0e4268f73b4e184d16f3b734f9e57/examples/hashmap/hashmap.c#L45)):*
<details><summary><i>Previously, DamianDuy (Damian Duy) wrote…</i></summary><blockquote>

Can those slashes be removed?
</blockquote></details>

I don't think so. They are needed for marking a new line when processing the `#define`.

**Review by lukaszstolarczuk — COMMENTED** [review]
> Reviewed 1 of 2 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/pmem/miniasync/71)* status: all files reviewed, 5 unresolved discussions (waiting on @DamianDuy and @kswiecicki)

___
*[`examples/hashmap/hashmap.c` line 454 at r12](https://reviewable.io/reviews/pmem/miniasync/71#-N1sJT3k7kni5b3BesAx:-N1sJT3k7kni5b3BesAy:ba62os3) ([raw file](https://github.com/pmem/miniasync/blob/f8dfa4efa4a0e4268f73b4e184d16f3b734f9e57/examples/hashmap/hashmap.c#L454)):*
> ```C
> 		struct hashmap_lookup_lock_entry_output);
> 
> /* Maps 'lookup' future entry data to the 'set_state' future entry data */
> ```

output data to ... entry data
___
*[`examples/hashmap/hashmap.c` line 596 at r12](https://reviewable.io/reviews/pmem/miniasync/71#-N1sK9Y74PaOv7hVgwUB:-N1sK9Y86LZQeEMq__P6:bb7p4x2) ([raw file](https://github.com/pmem/miniasync/blob/f8dfa4efa4a0e4268f73b4e184d16f3b734f9e57/examples/hashmap/hashmap.c#L596)):*
> ```C
> 	size_t buf_size = (size_t)arg;
> 
> 	/* We will not copy more data that buffer can fit */
> ```

than
___
*[`examples/hashmap/hashmap.c` line 816 at r12](https://reviewable.io/reviews/pmem/miniasync/71#-N1sM1E8-8E-hTbMMYOp:-N1sM1E8-8E-hTbMMYOq:b-96kdod) ([raw file](https://github.com/pmem/miniasync/blob/f8dfa4efa4a0e4268f73b4e184d16f3b734f9e57/examples/hashmap/hashmap.c#L816)):*
> ```C
> 
> /*
>  * Maps the 'entry_init' future entry data to the 'memcpy_value' future entry
> ```

.

output/result to ... data

**Review by kswiecicki — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/pmem/miniasync/71)* status: 2 of 4 files reviewed, 5 unresolved discussions (waiting on @DamianDuy, @lukaszstolarczuk, and @pbalcer)

___
*[`examples/hashmap/hashmap.c` line 454 at r12](https://reviewable.io/reviews/pmem/miniasync/71#-N1sJT3k7kni5b3BesAx:-N1sjDpw3d6ybx1AKyGo:b-896fix) ([raw file](https://github.com/pmem/miniasync/blob/f8dfa4efa4a0e4268f73b4e184d16f3b734f9e57/examples/hashmap/hashmap.c#L454)):*
<details><summary><i>Previously, lukaszstolarczuk (Łukasz Stolarczuk) wrote…</i></summary><blockquote>

output data to ... entry data
</blockquote></details>

Done.
___
*[`examples/hashmap/hashmap.c` line 596 at r12](https://reviewable.io/reviews/pmem/miniasync/71#-N1sK9Y74PaOv7hVgwUB:-N1sjE7x9TI-X4l4XbXn:b-896fix) ([raw file](https://github.com/pmem/miniasync/blob/f8dfa4efa4a0e4268f73b4e184d16f3b734f9e57/examples/hashmap/hashmap.c#L596)):*
<details><summary><i>Previously, lukaszstolarczuk (Łukasz Stolarczuk) wrote…</i></summary><blockquote>

than
</blockquote></details>

Done.
___
*[`examples/hashmap/hashmap.c` line 816 at r12](https://reviewable.io/reviews/pmem/miniasync/71#-N1sM1E8-8E-hTbMMYOp:-N1sjEXC3TQ9mubwv3jR:b-896fix) ([raw file](https://github.com/pmem/miniasync/blob/f8dfa4efa4a0e4268f73b4e184d16f3b734f9e57/examples/hashmap/hashmap.c#L816)):*
<details><summary><i>Previously, lukaszstolarczuk (Łukasz Stolarczuk) wrote…</i></summary><blockquote>

.

output/result to ... data
</blockquote></details>

Done.

**Review by lukaszstolarczuk — COMMENTED** [review]
> Reviewed 1 of 2 files at r13, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/pmem/miniasync/71)* status: 3 of 4 files reviewed, 2 unresolved discussions (waiting on @DamianDuy and @pbalcer)

**Review by DamianDuy — COMMENTED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/pmem/miniasync/71#-:-N2BB_9pC8ioFs714buw:bnfp4nl)* status: 3 of 4 files reviewed, 1 unresolved discussion (waiting on @pbalcer)

**Pull request merged**